I make a little experiment for this dataset and make a neural network classifying using scikit-learn library.

In [91]:
import pandas as pd

In [92]:
df = pd.read_csv('../input/AirQualityUCI.csv')

In [93]:
df.head()

Date      Time  CO_GT  PT08_S1_CO  NMHC_GT  C6H6_GT  PT08_S2_NMHC  \
0  11/23/2004  19:00:00   11.9        2008     -200     50.6          1980   
1  11/23/2004  20:00:00   11.5        1918     -200     49.4          1958   
2  11/17/2004  18:00:00   10.2        1802     -200     47.7          1924   
3  11/23/2004  18:00:00   10.2        1982     -200     49.5          1959   
4  11/26/2004  18:00:00   10.1        1956     -200     45.2          1877   

   Nox_GT  PT08_S3_Nox  NO2_GT  PT08_S4_NO2  PT08_S5_O3     T    RH      AH  \
0    1389          325     220         2562        2342  12.4  74.7  1.0741   
1    1358          335     190         2477        2237  11.5  76.2  1.0324   
2     748          356     192         2235        2452  13.7  52.8  0.8244   
3    1369          322     227         2536        2386  13.2  72.6  1.0936   
4    1389          347     255         2338        2465  15.5  62.8  1.0979   

    CO_level  
0  Very High  
1  Very High  
2  Very High  
3  Very High  
4  Very High

In [94]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9357 entries, 0 to 9356
Data columns (total 16 columns):
Date            9357 non-null object
Time            9357 non-null object
CO_GT           9357 non-null float64
PT08_S1_CO      9357 non-null int64
NMHC_GT         9357 non-null int64
C6H6_GT         9357 non-null float64
PT08_S2_NMHC    9357 non-null int64
Nox_GT          9357 non-null int64
PT08_S3_Nox     9357 non-null int64
NO2_GT          9357 non-null int64
PT08_S4_NO2     9357 non-null int64
PT08_S5_O3      9357 non-null int64
T               9357 non-null float64
RH              9357 non-null float64
AH              9357 non-null float64
CO_level        9357 non-null object
dtypes: float64(5), int64(8), object(3)
memory usage: 1.1+ MB


I do some data preprocessing like Data Cleaning and Data Reduction. <br>
1. Data Reduction<br>
The dataset have 16 columns and 9357 rows. I choose 13 columns for data mining. I'm not using date, time, and C6H6_GT attributes. You can see description about the attributes, and although maybe this dataset for timeseries, but i'm not using date and time. Whereas C6H6_GT haven't sensor measure result, since it's not relevant with another attributes that have actual and sensor measures. But, this is just my opinion.<br>
2. Data Cleaning<br>
There are some missing value in this dataset, you can see there are 200 and -200 values in each column, that's missing value. This missing value can make a bad classify result because the value sometimes not relevant with range in some columns. I fill the missing value with mean or median value from each columns. But, first i check if it's a null value.

In [95]:
#reduce columns
df = df.drop('Date', axis=1)
df = df.drop('Time', axis=1)
df = df.drop('C6H6_GT', axis=1)

In [96]:
#check null value
df.isnull().sum()

CO_GT           0
PT08_S1_CO      0
NMHC_GT         0
PT08_S2_NMHC    0
Nox_GT          0
PT08_S3_Nox     0
NO2_GT          0
PT08_S4_NO2     0
PT08_S5_O3      0
T               0
RH              0
AH              0
CO_level        0
dtype: int64

Seems like no null value in data. Let's check the missing values.

In [15]:
df['CO_GT'].min()

-200.0

In [17]:
df['PT08_S1_CO'].min()

-200

In [18]:
df['NMHC_GT'].min()

-200

In [19]:
df['PT08_S2_NMHC'].min()

-200

In [20]:
df['Nox_GT'].min()

-200

In [21]:
df['PT08_S3_Nox'].min()

-200

In [22]:
df['NO2_GT'].min()

-200

In [23]:
df['PT08_S4_NO2'].min()

-200

In [24]:
df['PT08_S5_O3'].min()

-200

In [25]:
df['T'].min()

-200.0

In [26]:
df['RH'].min()

-200.0

In [27]:
df['AH'].min()

-200.0

I do some cross check that every attributes imposibble have minus value except T (Temperature). But, it's impossible too if the Temperature -200, because this case about Very High, High, Moderate, Low, Very Low pollution in the air. <br>
But first, i fill this missing value with 0 value, because if i fill with mean, the value was not representative because it's contain with -200 value.

In [144]:
df.CO_level.unique()

array(['Very High', 'High', 'Moderate', 'Low', 'Very low'], dtype=object)

In [97]:
for i in df.ix[df['CO_GT']<0].index:
    df.loc[i, 'CO_GT'] = 0
for i in df.ix[df['PT08_S1_CO']<0].index:
    df.loc[i, 'PT08_S1_CO'] = 0
for i in df.ix[df['NMHC_GT']<0].index:
    df.loc[i, 'NMHC_GT'] = 0
for i in df.ix[df['PT08_S2_NMHC']<0].index:
    df.loc[i, 'PT08_S2_NMHC'] = 0
for i in df.ix[df['Nox_GT']<0].index:
    df.loc[i, 'Nox_GT'] = 0
for i in df.ix[df['PT08_S3_Nox']<0].index:
    df.loc[i, 'PT08_S3_Nox'] = 0
for i in df.ix[df['NO2_GT']<0].index:
    df.loc[i, 'NO2_GT'] = 0
for i in df.ix[df['PT08_S4_NO2']<0].index:
    df.loc[i, 'PT08_S4_NO2'] = 0
for i in df.ix[df['PT08_S5_O3']<0].index:
    df.loc[i, 'PT08_S5_O3'] = 0
for i in df.ix[df['RH']<0].index:
    df.loc[i, 'RH'] = 0
for i in df.ix[df['AH']<0].index:
    df.loc[i, 'AH'] = 0
#this case because it's possible the Temperature have minus degree,
#so i just fill 0 that have -200 value
for i in df.ix[df['T']==-200].index:
    df.loc[i, 'T'] = 0

In [30]:
#if you check min value has change
df['CO_GT'].min()

0.0

In [32]:
df['CO_GT'].mean()
#1.765

1.7655445121299898

In [33]:
df['CO_GT'].median()
#1.5

1.5

In [34]:
df['PT08_S1_CO'].mean()
#1056.81

1056.8130811157423

In [35]:
df['PT08_S1_CO'].median()
#1053.0


1053.0

In [36]:
df['NMHC_GT'].mean()
#21.373

21.373730896654912

In [37]:
df['NMHC_GT'].median()
#0

0.0

In [38]:
df['PT08_S2_NMHC'].mean()
#902.41

902.4182964625414

In [39]:
df['PT08_S2_NMHC'].median()
#895.0

895.0

In [40]:
df['Nox_GT'].mean()
#203.649

203.6495671689644

In [41]:
df['Nox_GT'].median()
#141.0

141.0

In [42]:
df['PT08_S3_Nox'].mean()
#802.813

802.8131879876029

In [43]:
df['PT08_S3_Nox'].median()
#794.0

794.0

In [44]:
df['NO2_GT'].mean()
#93.245

93.24559153574863

In [45]:
df['NO2_GT'].median()
#96.0

96.0

In [46]:
df['PT08_S4_NO2'].mean()
#1399.302

1399.3026611093298

In [47]:
df['PT08_S4_NO2'].median()
#1446.0

1446.0

In [48]:
df['PT08_S5_O3'].mean()
#982.895

982.8950518328525

In [49]:
df['PT08_S5_O3'].median()
#942.0

942.0

In [50]:
df['T'].mean()
#17.601

17.601325211071927

In [51]:
df['T'].median()
#17.2

17.2

In [52]:
df['RH'].mean()
#47.308

47.308400128246056

In [53]:
df['RH'].median()
#48.6

48.6

In [54]:
df['AH'].mean()
#0.985

0.9854165544512143

In [55]:
df['AH'].median()
#0.9768

0.9768

You can see that NMHC_GT median value is 0, that's make posibble if NMHC_GT value contains many 0 values. so that's why i fill the missing value with mean value. But, in some cases like PT08_S4_NO2 and PT08_S5_O3 that have difference quite large between mean and median, possibility because both of this column have a lot of pretty high value. So, i choose to fill both of this attributes with median value.

In [135]:
for i in df.ix[df['CO_GT']==0].index:
    df.loc[i, 'CO_GT'] = df['CO_GT'].mean()
for i in df.ix[df['PT08_S1_CO']==0].index:
    df.loc[i, 'PT08_S1_CO'] = df['PT08_S1_CO'].mean()
for i in df.ix[df['NMHC_GT']==0].index:
    df.loc[i, 'NMHC_GT'] = df['NMHC_GT'].mean()
for i in df.ix[df['PT08_S2_NMHC']==0].index:
    df.loc[i, 'PT08_S2_NMHC'] = df['PT08_S2_NMHC'].mean()
for i in df.ix[df['Nox_GT']==0].index:
    df.loc[i, 'Nox_GT'] = df['Nox_GT'].median()
for i in df.ix[df['PT08_S3_Nox']==0].index:
    df.loc[i, 'PT08_S3_Nox'] = df['PT08_S3_Nox'].mean()
for i in df.ix[df['NO2_GT']==0].index:
    df.loc[i, 'NO2_GT'] = df['NO2_GT'].mean()
for i in df.ix[df['PT08_S4_NO2']==0].index:
    df.loc[i, 'PT08_S4_NO2'] = df['PT08_S4_NO2'].median()
for i in df.ix[df['PT08_S5_O3']==0].index:
    df.loc[i, 'PT08_S5_O3'] = df['PT08_S5_O3'].median()
for i in df.ix[df['T']==0].index:
    df.loc[i, 'T'] = df['T'].mean()
for i in df.ix[df['RH']==0].index:
    df.loc[i, 'RH'] = df['RH'].mean()
for i in df.ix[df['AH']==0].index:
    df.loc[i, 'AH'] = df['AH'].mean()

In [136]:
#and if you check minimum value has change
df['CO_GT'].min()

0.10000000000000001

In [137]:
X = df.drop('CO_level', axis=1)
y = df.loc[:,'CO_level'].values

In [145]:
X

CO_GT   PT08_S1_CO     NMHC_GT  PT08_S2_NMHC  Nox_GT  PT08_S3_Nox  \
0     11.900000  2008.000000   21.373731   1980.000000  1389.0   325.000000   
1     11.500000  1918.000000   21.376015   1958.000000  1358.0   335.000000   
2     10.200000  1802.000000   21.378300   1924.000000   748.0   356.000000   
3     10.200000  1982.000000   21.380584   1959.000000  1369.0   322.000000   
4     10.100000  1956.000000   21.382869   1877.000000  1389.0   347.000000   
5      9.900000  1881.000000   21.385155   1983.000000  1479.0   334.000000   
6      9.500000  1908.000000   21.387440   2007.000000   952.0   325.000000   
7      9.400000  1816.000000   21.389726   1851.000000  1184.0   341.000000   
8      9.300000  1056.813081   21.392012    902.418296  1310.0   802.813188   
9      9.200000  1778.000000   21.394298   1935.000000   859.0   349.000000   
10     9.100000  1056.926025   21.396584    902.514740  1253.0   802.898986   
11     9.100000  1701.000000   21.398871   1692.000000  1220.0   402.000000   
12     8.700000  1882.000000   21.401158   1920.000000  1014.0   328.000000   
13     8.700000  1704.000000   21.403445   1689.000000  1070.0   402.000000   
14     8.700000  1615.000000   21.405733   1712.000000  1230.0   384.000000   
15     8.600000  1510.000000   21.408020   1501.000000   881.0   456.000000   
16     8.600000  1057.038980   21.410308    902.611193  1235.0   802.984793   
17     8.600000  1057.151948   21.412596    902.707657  1237.0   803.070610   
18     8.600000  1683.000000   21.414885   1831.000000  1216.0   374.000000   
19     8.500000  1630.000000   21.417173   1622.000000  1089.0   421.000000   
20     8.400000  1808.000000   21.419462   1771.000000   686.0   373.000000   
21     8.400000  1824.000000   21.421751   1862.000000   860.0   340.000000   
22     8.400000  1822.000000   21.424041   1890.000000   918.0   347.000000   
23     8.400000  1669.000000   21.426330   1686.000000  1074.0   384.000000   
24     8.400000  1057.264928   21.428620    902.804131  1028.0   803.156436   
25     8.300000  1623.000000   21.430910   1547.000000   832.0   414.000000   
26     8.300000  1693.000000   21.433201   1590.000000  1189.0   412.000000   
27     8.100000  1961.000000  618.000000   1701.000000   478.0   537.000000   
28     8.100000  1870.000000   21.435491   1871.000000  1001.0   351.000000   
29     8.100000  1759.000000   21.437782   1588.000000  1155.0   437.000000   
...         ...          ...         ...           ...     ...          ...   
9327   2.106671   898.000000   52.521760    624.000000   129.0   987.000000   
9328   2.106896  1098.849801   52.527373    938.313674    88.0   834.746586   
9329   2.107121  1154.000000   52.532987    788.000000   223.0   709.000000   
9330   2.107346   802.000000   52.538601    440.000000    31.0  1413.000000   
9331   2.107572   874.000000   52.544216    472.000000    49.0  1179.000000   
9332   2.107797   879.000000   52.549832    447.000000    38.0  1317.000000   
9333   2.108022   884.000000   52.555448    528.000000    96.0  1148.000000   
9334   2.108247   980.000000   52.561064    744.000000   174.0   840.000000   
9335   2.108473   790.000000   52.566682    471.000000    81.0  1302.000000   
9336   2.108698  1009.000000   52.572300    710.000000   191.0   819.000000   
9337   2.108923   872.000000   52.577918    558.000000   114.0  1054.000000   
9338   2.109149   942.000000   52.583537    665.000000   134.0   923.000000   
9339   2.109374  1015.000000   52.589157    608.000000   113.0   873.000000   
9340   2.109600   979.000000   52.594777    547.000000    73.0   952.000000   
9341   2.109825   972.000000   52.600398    542.000000   103.0  1029.000000   
9342   2.110051   992.000000   52.606020    565.000000   121.0   912.000000   
9343   2.110276   940.000000   52.611642    583.000000    93.0   942.000000   
9344   2.110502   963.000000   52.617264    617.000000    85.0   902.000000   
9345   2.110727   971.000000   52.622888    585.000000    79

In [139]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
#transform shape of y
y = le.fit_transform(y)

In [140]:
from sklearn.cross_validation import train_test_split
#split data with 70%, 30%
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3,  random_state =0)

In [141]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
sc.fit(X_train)
X_train_std = sc.transform(X_train)
X_test_std = sc.transform(X_test)

In [142]:
#use nn for classify
from sklearn.linear_model import Perceptron
ppn =  Perceptron(n_iter=1000, eta0=0.01, random_state=0)
ppn.fit(X_train_std, y_train)

Perceptron(alpha=0.0001, class_weight=None, eta0=0.01, fit_intercept=True,
      n_iter=1000, n_jobs=1, penalty=None, random_state=0, shuffle=True,
      verbose=0, warm_start=False)

In [143]:
y_pred = ppn.predict(X_test_std)
print('Misclassified sample : %d' % (y_test != y_pred).sum())

from sklearn.metrics import accuracy_score
print('accuracy : %.2f' % accuracy_score(y_test, y_pred))

Misclassified sample : 578
accuracy : 0.79


But it's not desire that the accuracy not very high, maybe i should get better in preprocessing data.